keps/sig-storage/1432-volume-health-monitor/README.md

```diff
@@ -136,11 +137,52 @@ Two main parts are involved here in the architecture.
   - Kubelet already collects volume stats from CSI node plugin by calling CSI function NodeGetVolumeStats.
   - In addition to existing volume stats collected already, Kubelet will also check volume condition collected from the same CSI function and log events to Pods if volume condition is abnormal.
   - Note that currently we do not have CSI support for local storage. When the support is available, we will implement relavant CSI monitoring interfaces as well.
+  - Expose Volume Health information as Kubelet VolumeStats Metrics.
 
 The volume health monitoring by Kubelet will be controlled by a new feature gate called `VolumeHealth`.
 
 ## Implementation
 
+### Kubelet Metrics changes
+
+Add a new field in the [VolumeStats metrics API](https://github.com/kubernetes/kubernetes/blob/v1.22.1/staging/src/k8s.io/kubelet/pkg/apis/stats/v1alpha1/types.go#L263).
+
+```
+// VolumeStats contains data about Volume filesystem usage.
+type VolumeStats struct {
+	// Embedded FsStats
+	FsStats `json:",inline"`
+	// Name is the name given to the Volume
+	// +optional
+	Name string `json:"name,omitempty"`
+	// Reference to the PVC, if one exists
+	// +optional
+	PVCRef *PVCReference `json:"pvcRef,omitempty"`
+
+	// Note: Add the following new field
+	// Normal volumes are available for use and operating optimally.
+	// An abnormal volume does not meet these criteria.
+	// +optional
+	Abnormal *bool `json:"abnormal,omitempty"`
```

Comment:
Do you think there would be more than 2 statuses in the future? Such as "healthy", "degraded", "failed"? We may want to consider following the pattern that we did for volume_operations* metrics where we have a [status field](https://github.com/kubernetes/kubernetes/blob/34fb61beba52d9e1ab3f044a7be8b9f49d99534e/pkg/volume/util/metrics.go#L54) that could be extended with more values in the future.

Reply:
The example in the [status field] (https://github.com/kubernetes/kubernetes/blob/34fb61beba52d9e1ab3f044a7be8b9f49d99534e/pkg/volume/util/metrics.go#L54) is actually a label which has to go with a real metric which is a float in this example.
Can we make it an array here? i.e., HealthStatus []bool

Reply:
Updated the KEP in include a VolumeHealthStats struct.

Reply:
I think it might be better to use single metric with different labels. Also, we should perhaps not update any of types being used for stats reporting because they are used elsewhere too (like cadvisor).